src/libraries/Common/src/Interop/Windows/Kernel32/Interop.CompareString.cs

```diff
@@ -0,0 +1,20 @@
+// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+
+using System;
+using System.Runtime.InteropServices;
+
+internal static partial class Interop
+{
+    internal static partial class Kernel32
+    {
+        [LibraryImport(Libraries.Kernel32, EntryPoint = "CompareStringW", SetLastError = true, StringMarshalling = StringMarshalling.Utf16)]
+        internal static partial int CompareString(
+            uint locale,
+            uint dwCmpFlags,
+            IntPtr lpString1,
+            int cchCount1,
+            IntPtr lpString2,
+            int cchCount2);
```

Comment:
It's more efficient to use `char*` and get rid of `Marshal` at call site. Should this be included in this PR?

Reply:
If I understand correctly, that would mean changing these `IntPtr`s to `char*` (and adding the unsafe modifier to the method sig, then changing the body of https://github.com/dotnet/runtime/blob/031204bd3464f4b57e2fc215072c0d6c50ceefa4/src/libraries/System.DirectoryServices/src/System/DirectoryServices/ActiveDirectory/Utils.cs#L1843 to something like the code below - is that correct?

```c#
        internal static int Compare(string? s1, string? s2, uint compareFlags)
        {
            // This code block was specifically written for handling string comparison
            // involving null strings. The unmanaged API "Interop.Kernel32.CompareString"
            // does not handle null strings elegantly.
            //
            // This method handles comparison of the specified strings
            // if and only if either one of the two strings or both are null.
            if (s1 == null || s2 == null)
            {
                return string.Compare(s1, s2);
            }

            int result = 0;
            int cchCount1 = 0;
            int cchCount2 = 0;

            cchCount1 = s1.Length;
            cchCount2 = s2.Length;

            unsafe
            {
                fixed (char* p1 = s1)
                fixed (char* p2 = s2)
                {
                    result = Interop.Kernel32.CompareString(LCID, compareFlags, p1, cchCount1, p2, cchCount2);
                }
            }

            if (result == 0)
            {
                throw ExceptionHelper.GetExceptionFromErrorCode(Marshal.GetLastPInvokeError());
            }

            return (result - 2); // to give the semantics of <0, ==0, >0
        }
```

Reply:
The `LibraryImport` will do the string->char* conversion for you. Just use `string` in the interop method and omit the `fixed` statement.

Reply:
Ah cool, thanks for the tip! In that case I think I've followed your recommendation in badbd6feb553924c89fcc3eaa47f0b849408046c (the `StringMarhsalling` attribute was already set).